lib/services/role.go

```diff
@@ -948,6 +948,42 @@ func (set RoleSet) EnumerateDatabaseUsers(database types.Database, extraUsers ..
 	return result
 }
 
+// EnumerateKubernetesUsersAndGroups works on a given role set to return a minimal list of
+// configured kubernetes_users and kubernetes_groups for a given kubernetes cluster.
+// The wildcard selector is ignored, since it is not allowed for kubernetes principals.
+func (set RoleSet) EnumerateKubernetesUsersAndGroups(kube types.KubeCluster) (EnumerationResult, EnumerationResult) {
+	kubeUsersResult := NewEnumerationResult()
+	kubeGroupsResult := NewEnumerationResult()
+
+	// gather logins for checking from the roles
+	// no need to check for wildcards
+	var kubeUsers []string
+	var kubeGroups []string
+	for _, role := range set {
+		kubeUsers = append(kubeUsers, role.GetKubeUsers(types.Allow)...)
+		kubeUsers = append(kubeUsers, role.GetKubeUsers(types.Deny)...)
+		kubeGroups = append(kubeGroups, role.GetKubeGroups(types.Allow)...)
+		kubeGroups = append(kubeGroups, role.GetKubeGroups(types.Deny)...)
+	}
+
+	kubeUsers = apiutils.Deduplicate(kubeUsers)
+	kubeGroups = apiutils.Deduplicate(kubeGroups)
+
+	// check each individual user against the cluster.
+	for _, user := range kubeUsers {
+		err := set.checkAccess(kube, AccessMFAParams{Verified: true}, NewKubeUserMatcher(user))
+		kubeUsersResult.allowedDeniedMap[user] = err == nil
```

Comment:
This ends up hiding errors that are not AccessDenied errors, and can show users/groups that are listed as allowed as denied due to some possibly-transient error when checking access. I can see this is also how it is done in `EnumerateServerLogins`, but perhaps that is also not correct? Should this enumerate function also return an error for the cases where we failed to check access, rather than reporting that as denied?